### PR TITLE
fix #145 doing indexing in configurable blocks

### LINKED ISSUE #145: make indexing smarter so it can use less memory
Current `FolderBasedIndexingService` implementation create all `Callable` instances in go, which uses a lot of memory.

It would be better to be able to index in blocks of arbitrary size.

### PR DESCRIPTION
**Pull request by natan-cox-cognizone**

### CODE REVIEW

**Review by michielrogissart-cognizone — APPROVED** [review]

File: cube/src/main/java/zone/cogni/asquare/cube/index/StaticIndexIndexingService.java [hunk 1]
```diff
@@ -70,14 +73,16 @@
    */
   private final Map<String, String> queryTemplateParameters;
 
-  public StaticIndexIndexingService(@Nonnull SpelService spelService,
+  public StaticIndexIndexingService(int indexBlockSize,
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/StaticIndexIndexingService.java:89]: isn't this used yet? if so we need a (deprecated) constructor with a default value

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+ */
+class IndexFolderUriReport {
+
+  static class CollectionFolderUriReport {
+
+    private final IndexFolderUriReport indexFolderUriReport;
+    private final CollectionFolder collectionFolder;
+    private final List<String> uris;
+
+    public CollectionFolderUriReport(IndexFolderUriReport indexFolderUriReport,
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java]: if class is only used "interally", you can make it non-static and find parent class instance with `CollectionFolderUriReport.this`

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Keeps a list of all URIs linked to each of the collections. Only URIs are stored to save memory for indexing.
+ * When pulling out blocks of URIs which are going to be indexed the memory should remain at a predictable level.
+ */
+class IndexFolderUriReport {
+
+  static class CollectionFolderUriReport {
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java]: Inner classes at end of class

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+      return collectionFolder;
+    }
+
+    public List<String> getUris() {
+      return Collections.unmodifiableList(uris);
+    }
+  }
+
+  private final IndexFolder indexFolder;
+  private final List<CollectionFolderUriReport> collectionFolderUriReports = new ArrayList<>();
```
Comment (domivds) [inline cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java:14]: not threadsafe (more a global comment, to be checked where needed as indexing runs threaded)

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java (hunk 2 above)
Comment (natan-cox-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java]: true. still on the fence if I want to keep it in or move it out though.

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java (hunk 3 above)
Comment (natan-cox-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java]: yes sir!

File: cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java (hunk 4 above)
Comment (natan-cox-cognizone) [reply cube/src/main/java/zone/cogni/asquare/cube/index/IndexFolderUriReport.java:14]: this runs before we invoke the actual indexing into the threadpool

**Pull request merged**